tdp_server/api/dependencies.py

```diff
@@ -33,31 +33,29 @@
     },
 }
 
-DAG = Dag(settings.TDP_COLLECTIONS)
-EXECUTOR = AnsibleExecutor(settings.TDP_RUN_DIRECTORY)
-RUNNER_SERVICE = RunnerService(
-    DAG,
-    settings.TDP_RUN_DIRECTORY,
-    settings.TDP_VARS,
-)
-CLUSTER_VARIABLES = ClusterVariables.get_cluster_variables(settings.TDP_VARS)
-
 
 def get_db() -> Generator:
     with SessionLocal() as db:
         yield db
 
 
+@lru_cache()
 def get_dag() -> Dag:
-    return DAG
+    return Dag(settings.TDP_COLLECTIONS)
 
 
+@lru_cache()
 def get_cluster_variables() -> ClusterVariables:
-    return CLUSTER_VARIABLES
+    return ClusterVariables.get_cluster_variables(settings.TDP_VARS)
 
 
-def get_runner_service() -> RunnerService:
-    return RUNNER_SERVICE
+@lru_cache()
+def get_runner_service(dag: Dag = Depends(get_dag)) -> RunnerService:
```

Comment:
To use `@lru_cache()`, argument must be hashable, a `Dag` is hashable ?

Reply:
It's the function arguments that must be hashable, not the result

Reply:
Yes and the function argument is an instance of `Dag` no ?

Reply:
I think what you want is a singleton not a cache or maybe I don't understand the code.

Reply:
There's no implementation of singleton in python. `lru_cache` for this kind of behavior is the accepted practice.

My goal was to defer `Dag`, `ClusterVariables`, and `RunnerService` instanciation to be able to replace them easily in the tests.

Reply:
Yeah, from what I research on the internet, the default `hash` value is `id(X)//16` (id being the memory address)